### PR TITLE
Run node/etcd over https

### PR DESCRIPTION
**Pull request by liggitt**
- [x] osc build-logs using https
- [x] osc log using https
- [x] proxy to node using https
- [x] osc exec
- [x] update create-node-config to specify clientCA, update help

Follow-ups:

- [ ] Separate listen args for etcd and node to allow starting them in http-mode when the master is in https mode

### CODE REVIEW

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
@rajatchopra can you look at the commit with the vagrant sdn changes to pass in etcd certs?

**Comment by rajatchopra** [comment]
vagrant sdn changes look good.
Need to try it out manually though. openshift-sdn itself has not been tested much with etcd certs.

**Comment by liggitt** [comment]
@deads2k review config/create-node-config bits
@smarterclayton review buildlogs change

**Comment by smarterclayton** [comment]
For the name

**Comment by liggitt** [comment]
done

**Comment by TomasTomecek** [comment]
@liggitt 
What will be the new way to get logs through API?

**Comment by liggitt** [comment]
@TomasTomecek build logs would be retrieved the same way pod logs are, by proxying through the API server to the node. This doesn't change anything with that, except that you have to go through the API rather than directly to the node.

**Comment by liggitt** [comment]
In your code, you would stream from this URL rather than redirecting and contacting the node yourself: `self._build_url("proxy/buildLogs/%s" % build_id)`

**Comment by liggitt** [comment]
The direction we're moving for things like build logs is a subresource of the build itself, so you could stream this URL directly: `/namespaces/[ns]/builds/[build]/logs`, but we're not quite there yet

**Comment by TomasTomecek** [comment]
> In your code, you would stream from this URL rather than redirecting and contacting the node yourself: self._build_url("proxy/buildLogs/%s" % build_id)

Thanks. Will try.

> The direction we're moving for things like build logs is a subresource of the build itself, so you could stream this URL directly: `/namespaces/[ns]/builds/[build]/logs`, but we're not quite there yet

That would be awesome! Is there a trello card/PR/issue for that?

File: pkg/cmd/server/start/master_args.go [hunk 1]
```diff
@@ -219,7 +230,13 @@ func (args MasterArgs) BuildSerializeableEtcdConfig() (*configapi.EtcdConfig, er
 		StorageDir:    args.EtcdDir,
 	}
 
+	if args.ListenArg.UseTLS() {
```
Comment (deads2k) [inline pkg/cmd/server/start/master_args.go]: This should be based on whether etcd's bind address scheme is https.

Comment (liggitt) [inline pkg/cmd/server/start/master_args.go]: GetEtcdBindAddress derives from ListenArg...

**Comment by liggitt** [comment]
rebased, tweaked param name for client ca in create-node-config

File: pkg/cmd/server/start/master_args.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/server/start/master_args.go]: > GetEtcdBindAddress derives from ListenArg...

So.... I need to make sure that etcdAddr matches what will be produced by twisting up the listenAddr for my master?  I really want to be free of this, but do we need to try to figure out which configurations we want to support from the command line and then come up with some sane arguments to describe them?

File: pkg/cmd/server/api/validation/master.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+		_, urlErrs := ValidateURL(u, fmt.Sprintf("urls[%d]", i))
+		if len(urlErrs) > 0 {
+			allErrs = append(allErrs, urlErrs...)
+		}
+	}
+
+	if len(config.CA) > 0 {
+		allErrs = append(allErrs, ValidateFile(config.CA, "ca")...)
+	}
+	allErrs = append(allErrs, ValidateCertInfo(config.ClientCert, false)...)
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/master.go]: Shouldn't (probably can't here) use a cert without a CA.

File: pkg/cmd/server/api/validation/master.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+		// TODO warn when the CORS list does not include the assetConfig.publicURL host:port
+		// only warn cause they could handle CORS headers themselves in a proxy
+	}
+
+	if config.DNSConfig != nil {
+		allErrs = append(allErrs, ValidateHostPort(config.DNSConfig.BindAddress, "bindAddress").Prefix("dnsConfig")...)
+	}
+
+	allErrs = append(allErrs, ValidateEtcdConnectionInfo(config.EtcdClientInfo).Prefix("etcdClientInfo")...)
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/master.go]: Given both etcdConnectionInformation and etcdConfig, it is possible to know if they match.  Starting in process and then choosing to expose through a proxy is probably insane.

File: pkg/cmd/server/api/validation/master.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+	if len(config.CA) > 0 {
+		allErrs = append(allErrs, ValidateFile(config.CA, "ca")...)
+	}
+	allErrs = append(allErrs, ValidateCertInfo(config.ClientCert, false)...)
+
+	return allErrs
+}
+
+func ValidateEtcdConfig(config *api.EtcdConfig) errs.ValidationErrorList {
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/master.go]: Checks for peer and master addresses?  If present, we expect certain formats.

File: pkg/cmd/server/api/validation/master.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+	allErrs = append(allErrs, ValidateCertInfo(config.ClientCert, false)...)
+
+	return allErrs
+}
+
+func ValidateKubernetesMasterConfig(config *api.KubernetesMasterConfig) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if len(config.MasterIP) > 0 {
+		allErrs = append(allErrs, ValidateSpecifiedIP(config.MasterIP, "masterIP")...)
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/master.go]: Not directly related, but you moved it so it caught my eye.  In the interest of fully specifying, should we simply fail on empty MasterIP?  I'm for it.

**Comment by deads2k** [comment]
Nothing significant.  A few questions, not all of them directly related to this pull.

Otherwise, lgtm

File: pkg/cmd/server/api/validation/master.go (hunk 2 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/master.go]: If the etcd server is using a cert signed by a recognized authority, you don't need a custom CA

File: pkg/cmd/server/api/validation/master.go (hunk 4 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/master.go]: added

File: pkg/cmd/server/start/master_args.go (hunk 1 above)
Comment (liggitt) [inline pkg/cmd/server/start/master_args.go]: from args, you get no options for etcd server addresses... it uses the scheme from --listen, the host from --master, and port 4001. I'll add a follow up item to split listen args for etcd and node if we think it's important

File: pkg/cmd/server/api/validation/master.go (hunk 5 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/master.go]: I don't want to be more restrictive than kubernetes is, and that would mean you can't run Openshift using DNS only, which we may want

Comment (smarterclayton) [inline pkg/cmd/server/api/validation/master.go]: I think running DNS only is a very reasonable thing for upstream eventually.  Just not there yet.

----- Original Message -----

**Comment by liggitt** [comment]
Rebased, tests updated, comments addressed

**Comment by liggitt** [comment]
[merge]

**Comment by liggitt** [comment]
@sdodson @detiber  you'll want to add --node-client-certificate-authority to the create-node-config call

**Pull request merged**